### ISSUE TITLE
fsdp2 based hsdp, caculate grad_norm error in `clip_grad_norm`

### ISSUE THREAD

**Issue by ValMystletainn** [ByteDance-Seed/VeOmni#184]
### System Info

veomni 0.1.3
ubuntu 20.04
python 3.11.4
cuda 12.6

### Information

- [x] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

the grad norm is caculate by sum the p-th power of parameter elements, first in local than all-reduce with sum in `fsdp_group`, in current version, it is `dp_sp` group. However I think it should only do in `dp_shard` group.

It cause the grad norm mismatch when use large dp_replicate size, in my training (a from scratch model). and I just write a minimum demo scripts that can reproduce the norm error with only a node with 8 gpu.

```python
# reproduce_hsdp_clip_grad.py

from dataclasses import dataclass, field

import torch
import torch.distributed as dist

from veomni.distributed.parallel_state import init_parallel_state
from veomni.distributed.torch_parallelize import build_parallelize_model
from veomni.utils import helper
from veomni.utils.arguments import TrainingArguments, parse_args
from veomni.utils.device import get_device_type, get_nccl_backend, get_torch_device


logger = helper.create_logger(__name__)

@dataclass
class Argument:
    train: "TrainingArguments" = field(default_factory=TrainingArguments)

class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bias = torch.nn.Parameter(torch.ones(16), requires_grad=True)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        loss = (x + self.bias).sum()
        return loss

    def init_weights(self):
        self.bias.data.fill_(1.0)

def main():
    dist.init_process_group(backend=get_nccl_backend())
    args = parse_args(Argument)

    get_torch_device().set_device(f"{get_device_type()}:{args.train.local_rank}")
    init_parallel_state(
        dp_size=args.train.data_parallel_size,
        dp_replicate_size=args.train.data_parallel_replicate_size,
        dp_shard_size=args.train.data_parallel_shard_size,
        tp_size=args.train.tensor_parallel_size,
        ep_size=args.train.expert_parallel_size,
        pp_size=args.train.pipeline_parallel_size,
        cp_size=args.train.context_parallel_size,
        ulysses_size=args.train.ulysses_parallel_size,
        dp_mode=args.train.data_parallel_mode,
    )

    model = Model()
    model = build_parallelize_model(
        model,
        init_device=args.train.init_device,
        weights_path=None,
        enable_full_shard=args.train.enable_full_shard,
        enable_mixed_precision=args.train.enable_mixed_precision,
        enable_gradient_checkpointing=args.train.enable_gradient_checkpointing,
        enable_fsdp_offload=args.train.enable_fsdp_offload,
        basic_modules=[],
        enable_reentrant=args.train.enable_reentrant,
        enable_forward_prefetch=args.train.enable_forward_prefetch,
    )

    for step in range(10):
        inputs = torch.randn(1, 16).cuda()
        loss = model(inputs)
        loss.backward()
        _gn = model.clip_grad_norm_(args.train.max_grad_norm)
        grad_norm = _gn.item() if hasattr(_gn, "item") else float(_gn)
        logger.info_rank0(f"step: {step}, loss: {loss.item()}, grad_norm: {grad_norm}")
        model.zero_grad()

    dist.barrier()
    dist.destroy_process_group()

if __name__ == "__main__":
    main()

```
since this is a toy model. we can calculate the norm first with math. the correct grad norm should be $\sqrt{1^2 * 16} = 4$

running results

1. full shard: the grad norm is correct
```bash
torchrun --standalone --nproc-per-node 8 \
  reproduce_hsdp_clip_grad.py \
  --train.data_parallel_shard_size 8 \
  --train.data_parallel_mode fsdp2 \
  --train.init_device meta \
  --train.output_dir "debug"
```

the log is
```
11/10/2025 04:56:01 - INFO - __main__ - step: 0, loss: 21.6999454498291, grad_norm: 3.999999761581421
11/10/2025 04:56:01 - INFO - __main__ - step: 1, loss: 12.907739639282227, grad_norm: 3.999999761581421
11/10/2025 04:56:01 - INFO - __main__ - step: 2, loss: 17.441078186035156, grad_norm: 3.999999761581421
11/10/2025 04:56:01 - INFO - __main__ - step: 3, loss: 12.981290817260742, grad_norm: 3.999999761581421
11/10/2025 04:56:01 - INFO - __main__ - step: 4, loss: 21.56136703491211, grad_norm: 3.999999761581421
11/10/2025 04:56:01 - INFO - __main__ - step: 5, loss: 17.38682746887207, grad_norm: 3.999999761581421
...
```

2. mock hybird shard with 4 replicate with shard size = 2

```bash
torchrun --standalone --nproc-per-node 8 \
  reproduce_hsdp_clip_grad.py \
  --train.data_parallel_shard_size 2 \
  --train.data_parallel_mode fsdp2 \
  --train.init_device meta \
  --train.output_dir "debug"
```

the log is 

```
11/10/2025 05:08:50 - INFO - __main__ - step: 0, loss: 22.36000633239746, grad_norm: 7.999999523162842
11/10/2025 05:08:50 - INFO - __main__ - step: 1, loss: 21.73548698425293, grad_norm: 7.999999523162842
11/10/2025 05:08:50 - INFO - __main__ - step: 2, loss: 17.987272262573242, grad_norm: 7.999999523162842
11/10/2025 05:08:50 - INFO - __main__ - step: 3, loss: 11.49040699005127, grad_norm: 7.999999523162842
11/10/2025 05:08:50 - INFO - __main__ - step: 4, loss: 13.00349235534668, grad_norm: 7.999999523162842
11/10/2025 05:08:50 - INFO - __main__ - step: 5, loss: 17.158998489379883, grad_norm: 7.999999523162842
...
```

it's doubled. it is caused the powered sum of parameter grads is sum in error group, turn's to $\sqrt{4 * 2 * 8 * 1^2}$  (replicate sum, shard sum, local sum)

 

### Expected behavior

the grad norm of the two demo runs in **Reproduction** should both be approxmaitly `4`

I think in `veomni/distributed/fsdp2/clip_grad_norm.py`
we should turn's the group that gives `_fsdp2_reduce_and_clip` from `fsdp_group` to `dp_shard_group`

**Comment by Luosuu** [comment]
Thank you very much for the bug report! This is a good opportunity to refactor the grad norm clipping implementation for FSDP2.. IMO, it should be possible to use the torch native clipping instead of manually reducing (except for EP+FSDP2 which is a bit more complicated). I will fix it soon

**Comment by Luosuu** [comment]
Hi I just made a PR and hopefully it will be merged soon (sorry that our team is a bit busy now)

**Comment by Luosuu** [comment]
the fix PR has been merged. Feel free to reopen if the issue persists
